added tutorial in new branch (#968)
Qiskit/qiskit-experiments: qiskit_experiments/test/Fine Calibration Tutorial.rst

================================================

> Fine Calibrations with Pulse Simulator

Reviewer: ```suggestion
Fine calibration of a pulse amplitude
```

================================================

> The amplitude of a pulse can be precisely calibrated using

Reviewer: ```suggestion
Calibrating quantum gates is the task of finding the parameters of the underlying pulses that best implement the target gate.
The amplitude of a pulse can be precisely calibrated using
```

error amplifying gate sequences. These gate sequences apply 
the same gate a variable number of times. Therefore, if each gate

> has a small error dtheta  in the rotation angle then 
> a sequence of n gates will have a rotation error of n*dtheta.

Reviewer: Can you use :math:? I think it should render properly when you build the docs.
```suggestion
has a small error :math:`{\rm d}\theta`  in the rotation angle then 
a sequence of :math:`n` gates will have a rotation error of :math:`n\cdot {\rm d}\theta`.
```

Author: done

Since our new pulse simulator cache the calibrations that it has already solved
in the form of dictionary in which keys are a tuple of instruction name, qubits and parameters
and the vaule is a unitary matrix, it is efficient when simulating a circuit
which consists of same repeated gates. 
Therefore, fine amplitude calibration experiment which is a typical example of
containing such repeated gates within a circuit, can be run efficiently in the "IQPulseBackend"

> Let's start with importing some necessary modules and build our pulse simulator with
> "SingleTransmonTestBackend" which is inherited from "IQPulseBackend"

Reviewer: ```suggestion
```

.. jupyter-execute:: 

    import numpy as np
    from qiskit.pulse import InstructionScheduleMap
    import qiskit.pulse as pulse
    from qiskit_experiments.library import FineXAmplitude, FineSXAmplitude

>     from qiskit_experiments.test.iq_pulse_backend import SingleTransmonTestBackend

Reviewer: ```suggestion
    from qiskit_experiments.test.pulse_backend import SingleTransmonTestBackend
```

.. jupyter-execute::

    backend = SingleTransmonTestBackend()
    qubit = 0
-----------------------------------------------------
Fine X gate Amplitude Calibration
-----------------------------------------------------
We will run the fine calibration experiments with our own pulse schedules. 
To do this we create an instruction to schedule map which we populate with 
the schedules we wish to work with. This instruction schedule map is then 

> given to the transpile options of the calibration experiments so that 
> the Qiskit transpiler can attach the pulse schedules to the gates in the experiments. 

Reviewer: ```suggestion
given to the transpile options of the experiment. 
```

> We will base all our pulses on the default X pulse of "SingleTransmonTestBackend".

Reviewer: ```suggestion
We base all our pulses on the default X pulse of "SingleTransmonTestBackend".
```

.. jupyter-execute::
    x_pulse = backend.defaults().instruction_schedule_map.get('x', (qubit,)).instructions[0][1].pulse

>     d0, inst_map = pulse.DriveChannel(qubit), InstructionScheduleMap()

Reviewer: ```suggestion
    d0, inst_map = pulse.DriveChannel(qubit), pulse.InstructionScheduleMap()
```

    for name, pulse_ in [("x", x_pulse)]:
        with pulse.build(name=name) as sched:
            pulse.play(pulse_, d0)

>         inst_map.add(name, (qubit,), sched)

Reviewer: This is a bit confusing. I would only add one pulse to the inst. map. Namely the one with the over or under rotation.

Reviewer: ```suggestion
```

We now take the ideal x pulse amplitude reported by the backend and 
add/subtract a 2% over/underrotation to it by scaling the ideal amplitude and see 
if the experiment can detect this over/underrotation. We replace the default X pulse 
in the instruction schedule map with this over/underrotated pulse.

.. jupyter-execute::
    ideal_amp = x_pulse.amp
    over_amp = ideal_amp*1.02
    under_amp = ideal_amp*0.98

>     print(f"The reported amplitude of the X pulse is {ideal_amp:.4f} which we set as ideal_amp. we use {over_amp:.4f} amplitude for overroation pulse and {under_amp:.4f} for underrotation pulse")

Reviewer: Can you break this line in two?

> -----------------------------------------------------------------------------------
> Detecting an over & under rotated pulse
> -----------------------------------------------------------------------------------

Reviewer: we can remove this title
```suggestion
```

.. jupyter-execute::
    # build the over rotated pulse and add it to the instruction schedule map
    with pulse.build(backend=backend, name="x") as x_over:
        pulse.play(pulse.Drag(x_pulse.duration, over_amp, x_pulse.sigma, x_pulse.beta), d0)

>     inst_map.add("x", (qubit,), x_over)

Reviewer: Can you move this to the previous code block where you set the pulse parameters?

    # do the experiment
    overamp_cal = FineXAmplitude(qubit, backend=backend)
    overamp_cal.set_traspile_options(inst_map=inst_map)
    exp_data_over = overamp_cal.run(backend)

>     exp_data_over.figure(0)

Reviewer: Can you make a cell that prints out one of the circuits? Perhaps one showing 3 X gates? Also make the comment that, for the calibration of the X gate we have a square-root of X gate before the X gates. This is there to move the ideal population to the equator of the Bloch sphere where the population change is most sensitive to over/under rotations.

Author: I'm confused with the purpose of square root X.
1. you said the rason for SX is the sensitivity of equator population change.
Q1. Isnt it right that the change in population independent of qubit intial state? Whether the qubit was in it's ground state or superposition state, isnt the over/under rotation amplitude affect the qubit in the same way?
2. In the qiskit-experiment tutorial, they say the reason for SX is to be able to distinguish between over-rotated and under-rotated pulses.
Q2. I think we can distinguish between the over and under rotated pulse even when qubit starts from the ground state. Isnt it??

> .. jupyter-execute::

Reviewer: It is good that you do both over and under rotation of the gates. However, you need to comment on the figures that are being shown. Something like saying that the ping-pong pattern that we see is flip depending on whether we have an under or an over rotation. Basically we need a small piece of text to explain to the reader what she/he is seeing.

    underamp_cal.set_traspile_options(inst_map=inst_map)
    exp_data_under = underamp_cal.run(backend)
    exp_data_under.figure(0)

.. jupyter-execute::
    # analyze the results
    target_angle = np.pi
    dtheta_over = exp_data_over.analysis_results("d_theta").value.nominal_value
    scale_over = target_angle / (target_angle + dtheta_over)
    dtheta_under = exp_data_under.analysis_results("d_theta").value.nominal_value
    scale_under = target_angle / (target_angle + dtheta_under)
    print(f"The ideal angle is {target_angle:.2f} rad. We measured a deviation of {dtheta_over:.3f} rad in over-rotated pulse case.")
    print(f"Thus, scale the {over_amp:.4f} pulse amplitude by {scale_over:.3f} to obtain {over_amp*scale_over:.5f}.")
    print(f"On the other hand, we measued a deviation of {dtheta_under:.3f} rad in under-rotated pulse case.")
    print(f"Thus, scale the {under_amp:.4f} pulse amplitude by {scale_under:.3f} to obtain {under_amp*scale_under:.5f}.")

-----------------------------------------------------------------------------------
Analyzing a pi/2 pulse
-----------------------------------------------------------------------------------

> .. jupyter-execute::

Reviewer: This is good but it needs some text. You point out some of the differences with calibrating an X gate. Namely, that with the SX gate we simply need to chose the right number of repetitions to always have the ideal population land on the equator of the Bloch sphere. Also, there should be on repetition done with two sx gates which allows us to calibrate the amplitude of the ping-pong pattern. This helps take measurement errors into account.

    exp_data_x90p = amp_cal.run().block_for_results()
    exp_data_x90p.figure(0)

.. jupyter-execute::

    # check how much more the given sx_pulse makes over or under roatation
    print(exp_data_x90p.analysis_results("d_theta"))
    target_angle = np.pi / 2
    dtheta = exp_data_x90p.analysis_results("d_theta").value.nominal_value
    scale = target_angle / (target_angle + dtheta)
    print(f"The ideal angle is {target_angle:.2f} rad. We measured a deviation of {dtheta:.3f} rad.")
    print(f"Thus, scale the {sx_pulse.amp:.4f} pulse amplitude by {scale:.3f} to obtain {sx_pulse.amp*scale:.5f}.")

Let's change the sx_pulse with the scaled sx_pulse expecting it to make a sharp pi/2 rotation.
(dtheta~0.000)

.. jupyter-execute::

    pulse_amp = sx_pulse.amp*scale

    with pulse.build(backend=pulse_backend, name="sx") as sx_new:
        pulse.play(pulse.Drag(x_pulse.duration, pulse_amp, x_pulse.sigma, x_pulse.beta), d0)

    inst_map.add("sx", (qubit,), sx_new)
    inst_map.get('sx',(qubit,))

.. jupyter-execute::

    # do the experiment
    data_x90p = amp_cal.run().block_for_results()
    data_x90p.figure(0)

.. jupyter-execute::

    # check the dtheta 

>     print(data_x90p.analysis_results("d_theta"))

Reviewer: What is this?

Author: When we do FineSXAmplitude experiment, the result shows inaccurate pulse amplitude like the first photo. So I added how to get the fine SX amplitude by multiplying the `scale` and applying the new_sx to get the second photo.
![image](https://user-images.githubusercontent.com/56623045/201524656-86035193-3988-4801-b805-6ef099c1f61f.png)
![image](https://user-images.githubusercontent.com/56623045/201524668-1b35de18-4ea1-480e-938c-118864df63e1.png)
